### PR TITLE
MeasureCriteria to be considered for indicator values

### PR DESCRIPTION
**Pull request by ameenhere**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java [hunk 1]
```diff
@@ -35,5 +35,43 @@
...
+    /**
+     * Checks the constraint using x and y as the values and uses the current
+     * measurefilter instance for comparison.
+     * 
+     * @param x The first double value to be compared
+     * @param y The second double value to be compared
+     * @return true if the constraint/filter is satisfied when x is compared
+     *         with y.
+     */
+    public boolean checkConstraint( Double x, Double y )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java]: Perhaps call `measureIsValid` or something that illustrates the returned method value?

Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java]: Yep. Sounds good.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java [hunk 2]
```diff
@@ -482,6 +484,33 @@ private void addIndicatorValues( DataQueryParams params, Grid grid )
...
+
+        // Prepare the roundedValue, as per the dataQueryParams, for comparison
+        Double indicatorRoundedValue = AnalyticsUtils.getRoundedValue( params, indicator.getDecimals(), value.getValue() );
+
+        // If any measureCriteria filter fails (returns false) when checking
+        // constraint, then return false. If no measureCriteria fails, then
+        // return true.
+        return !params.getMeasureCriteria().entrySet().stream().anyMatch( e -> {
+            return !e.getKey().checkConstraint( indicatorRoundedValue, e.getValue() );
+        } );
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java]: Yep. Changed it. 👍

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java [hunk 3]
```diff
@@ -482,6 +484,33 @@ private void addIndicatorValues( DataQueryParams params, Grid grid )
...
+    private boolean satisfiesMeasureCriteria( DataQueryParams params, IndicatorValue value, Indicator indicator )
+    {
+        // Skip measureCriteria check if there is no measure criteria
+        if ( !params.hasMeasureCriteria() )
+        {
+            return true;
+        }
+
+        // Prepare the roundedValue, as per the dataQueryParams, for comparison
+        Double indicatorRoundedValue = AnalyticsUtils.getRoundedValue( params, indicator.getDecimals(), value.getValue() );
```
[the remark this answers is not in the document]
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java]: Just a point here. Eg. if you get 2.98 and the criteria is <3.0 , the value will be included but shown as "3.0" in the UI. (Not strictly less than 3.0). Will this create a confusion for the user? One of the reason I did it this way to filter it based on what the user was going to see.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java [hunk 4]
```diff
@@ -482,6 +484,33 @@ private void addIndicatorValues( DataQueryParams params, Grid grid )
...
+
+        // If any one measureCriteria filter fails (returns false) during
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java]: Reduce comments to one line, fairly clear what's going on here

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java [hunk 5]
```diff
@@ -35,5 +35,42 @@
...
+
+            return Double.compare( x, y ) == 0;
+
+        case GT:
+
+            return Double.compare( x, y ) > 0;
+
+        case GE:
+
+            return Double.compare( x, y ) > 0 || Double.compare( x, y ) == 0;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java]: Double.compare( x, y ) >= 0   ?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java [hunk 6]
```diff
@@ -35,5 +35,42 @@
...
+
+        case LT:
+
+            return Double.compare( x, y ) < 0;
+
+        case LE:
+
+            return Double.compare( x, y ) < 0 || Double.compare( x, y ) == 0;
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java]: Double.compare( x, y ) <= 0    ?

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java (hunk 5 above)
Comment (ameenhere) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/MeasureFilter.java]: This is embarassing. 😆

**Pull request merged**